docs/user-defined-type-functions.md

```diff
@@ -0,0 +1,137 @@
+end
+
+type Person = {
+    name: string,
+    age: number
+}
+
+type ty = rawget<Person, "name"> -- ty = string
+```
+To allow Luau developers to modify the runtime values of types in type functions, this RFC proposes introducing a new userdata called `lType`.  An `lType` object is a runtime representation of all types within the program and provides a basic set of library methods that can be used to modify types. As such, under the hood, each argument of a type function is serialized into a userdata called `lType`. Moreover, they are *only accessible within type functions* and are *not a runtime type for other use cases than type functions*. 
+
+<details><summary>lType library methods (dropdown)</summary>
+Note: methods under a different type heading (ex: `Singleton`) imply that the methods are only available for those types. At the implementation level, there is a check to make sure that the type-specific methods are being called on the correct types (e.g, for `getIndexer()`, assert that `isTable()` is true).
+
+#### Any
+
+| Function Declaration | Return Type | Description |
+| ------------- | ------------- | ------------- |
+| `isString()` | `boolean` | returns true if self is of type `string` |
+| `isNumber()` | `boolean` | returns true if self is of type `number` |
+| `isBoolean()` | `boolean` | returns true if self is of type `boolean` |
+| `isTable()` | `boolean` | returns true if self is of type `table` |
+| `isThread()` | `boolean` | returns true if self is of type `thread` |
+| `isFunction()` | `boolean` | returns true if self is of type `function` |
+| `isBuffer()` | `boolean` | returns true if self is of type `buffer` |
+| `isNil()` | `boolean` | returns true if self is of type `nil` |
+| `isClass()` | `boolean` | returns true if self is of type `class` (do we need this?) |
+| `isBooleanSingleton()` | `boolean` | returns true if self is a boolean singleton |
+| `isStringSingleton()` | `boolean` | returns true if self is a string singleton |
+| `isA(arg: lType)` | `boolean` | returns true if arg is the same type as self |
+
+#### Primitive
+
+| Function Declaration | Return Type | Description |
+| ------------- | ------------- | ------------- |
+| `getType()` | `string` | returns either "nil", "boolean", "string", "thread", "function", "table", or "buffer" |
+
+#### Singleton
+
+| Function Declaration | Return Type | Description |
+| ------------- | ------------- | ------------- |
+| `getValue()` | `string` | returns either "true", "false", or a string singleton |
+
+#### Table
+
+| Function Declaration | Return Type | Description |
+| ------------- | ------------- | ------------- |
+| `getProps()` | `table` | returns a type representation of tables (e.g. {name = "John"} will return {[string] = "string"}) |
+| `getIndexer()` | `table` | returns a type representation of arrays (e.g. {1, "hi", 3} will return {[number] = "number" \| "string"}) |
+
+#### Class
```

Comment:
userdata have properties and indexers too, but this proposal doesn't currently give us a way to access them.